internal/workflow.go

```diff
@@ -46,92 +46,229 @@ var (
+	//
+	// Finally, note that Select will not return until a condition's needs are met, like a Go selector - canceling the
+	// Context used to construct the Selector, or the Context used to Select, will not (directly) unblock a Select call.
+	// Read Select for more details.
 	Selector interface {
-		// AddReceive adds a ReceiveChannel to the selector. f is invoked when the channel has data or closed.
-		// ok == false indicates the channel is closed
-		AddReceive(c Channel, f func(c Channel, ok bool)) Selector
-		// AddSend adds a SendChannel to the selector. f is invoke when the channel is available to send
+		// AddReceive waits to until a value can be received from a channel.
+		// f is invoked when the channel has data or is closed.
+		//
+		// This is equivalent to `case v, more := <- aChannel`, and `more` will only
+		// be false when the channel is both closed and no data was received.
+		//
+		// When f is invoked, the data (or closed state) remains untouched in the channel, so
+		// you need to `c.Receive(ctx, &out)` (or `c.ReceiveAsync(&out)`) to remove and decode the value.
+		// Failure to do this is not an error - the value will simply remain in the channel until a future
+		// Receive retrieves it.
+		AddReceive(c Channel, f func(c Channel, more bool)) Selector
+		// AddSend waits to send a value to a channel.
+		// f is invoked when the value was successfully sent to the channel.
+		//
+		// This is equivalent to `case aChannel <- value`.
+		//
+		// Unlike AddReceive, the value has already been sent on the channel when f is invoked.
 		AddSend(c Channel, v interface{}, f func()) Selector
-		// AddFuture adds a Future to the selector f is invoked when future is ready
+		// AddFuture invokes f after a Future is ready.
+		// If the Future is ready before Select is called, it is eligible to be invoked immediately.
+		//
+		// There is no direct equivalent in a native Go select statement.
+		// It was added because Futures are common in Cadence code, and some patterns are much simpler with it.
+		//
+		// Each call to AddFuture will invoke its f at most one time, regardless of how many times Select is called.
+		// This means, for a Future that is (or will be) ready:
+		// - Adding the Future once, then calling Select twice, will invoke the callback once with the first Select
+		//   call, and then wait for other Selector conditions in the second Select call (or block forever if there are
+		//   no other eligible conditions).
+		// - Adding the same Future twice, then calling Select twice, will invoke each callback once.
+		// - Adding the same Future to two different Selectors, then calling Select once on each Selector, will invoke
+		//   each Selector's callback once.
+		//
+		// Therefore, with a Future "f" that is or will become ready, this is an infinite loop that will consume as much
+		// CPU as possible:
+		//  for {
+		//  	workflow.NewSelector(ctx).AddFuture(f, func(f workflow.Future){}).Select(ctx)
```

Comment:
TIL! 

Looks like we should encourage putting only `Select(ctx)` into forloop?

Reply:
I dunno tbh.  with simple demos like this, yeah - I think it's somewhat safer to construct it once and select multiple times.

But _mutating the select inside a callback_ is somewhat common (e.g. starting a new activity and waiting on it too), and from what I've seen people have fewer problems with that kind of setup (and with complex selects in general) when they rebuild it from scratch each time.  
it forces them to track state outside the loop earlier on, which is a good thing - they can't _remove_ those futures once they've been handled, so they are fairly often eventually forced to rebuild it in the loop, and that's not always a simple change.

Reply:
another way to look at it I guess:
- one way risks an infinite loop
- one way risks a deadlock

I don't really think there's a "best" option there - regardless of which you choose, you _must_ do it right.